Content.Server/GameTicking/Rules/RevolutionariesRuleSystem.cs

```diff
@@ -0,0 +1,242 @@
+        ).ToList();
+
+        var prefList = new List<IPlayerSession>();
+
+        // Adds every player from the list that has Revolutionary Head checked in their preference list to the preflist list
+        foreach (var player in list)
+        {
+            if (!ev.Profiles.ContainsKey(player.UserId))
+            {
+                continue;
+            }
+            var profile = ev.Profiles[player.UserId];
+            if (profile.AntagPreferences.Contains(RevolutionaryHeadPrototypeId))
+            {
+                prefList.Add(player);
+            }
+        }
+
+        var numRevoHeads = MathHelper.Clamp(_aliveCommandHeads.Count / headsPerRevoHead, 1, maxRevoHeads);
+
+        for (var i = 0; i < numRevoHeads; i++)
+        {
+            IPlayerSession revoHead;
+            if(prefList.Count == 0)
+            {
+                if (list.Count == 0)
+                {
+                    Logger.InfoS("preset", "Insufficient preffered players ready to fill up with revo heads, stopping the selection.");
+                    break;
+                }
+                revoHead = _random.PickAndTake(list);
+                Logger.InfoS("preset", "Insufficient preferred revo heads, picking at random.");
+            }
+            else
+            {
+                revoHead = _random.PickAndTake(prefList);
+                list.Remove(revoHead);
+                Logger.InfoS("preset", "Selected a preferred revo head.");
+            }
+            var mind = revoHead.ContentData()?.Mind;
+            if (mind is null)
+            {
+                Logger.ErrorS("preset", "Failed getting mind for picked revo head.");
+                return;
+            }
+
+            var antagPrototype = _prototypeManager.Index<AntagPrototype>(RevolutionaryHeadPrototypeId);
+            var revoHeadRole = new TraitorRole(mind, antagPrototype);
+            mind.AddRole(revoHeadRole);
+            _aliveRevoHeads.Add(mind, true);
+            
+            if (mind.OwnedEntity is null)
+                return;
+            
+            var messageWrapper = Loc.GetString("chat-manager-server-wrap-message");
+            _stationSpawningSystem.EquipStartingGear(mind.OwnedEntity.Value, _prototypeManager.Index<StartingGearPrototype>("RevoHeadGear"), null);
+
+            if (mind.Session == null)
+                return;
+            
+            var inCharacterName = mind.CharacterName;
+            if (inCharacterName != null)
+                _revolutionHeadNames.Add(inCharacterName, mind.Session.Name);
+
+            _chatManager.ChatMessageToOne(Shared.Chat.ChatChannel.Server, Loc.GetString("revolution-head-role-greeting"),
+               messageWrapper, default, false, mind.Session.ConnectedClient, Color.Red);
+        }
+
+        SoundSystem.Play(_addedSound.GetSound(), Filter.Empty().AddWhere(s => ((IPlayerSession)s).Data.ContentData()?.Mind?.HasRole<TraitorRole>() ?? false), AudioParams.Default);
+        
+    }
```

Comment:
This should support being added midround tbh, also the filter is cooked here as well.